### ISSUE TITLE
Newrelic integration

### ISSUE THREAD

**Issue by mschae** [phoenixframework/phoenix#402]
Hey all,

I am close to putting my first phoenix app into production (yay!). Therefore I was wondering if any of you had any experience with using new relic to monitor phoenix apps. Maybe point me at a good starting point?

I found this: https://github.com/wooga/newrelic-erlang

Not sure if and how we can use it.

Thanks,
Michael

**Comment by thawatchai** [comment]
I'm new to both elixir and phoenix but successfully deployed a phoenix app to serve some api endpoints in our sites. This is to test the performance of phoenix. So far, it outperforms the existing node app!!

The next step for us is to monitor the phoenix app with New Relic before we port more parts to phoenix. If anyone could expand @chrismccord 's answer, it would be really helpful for new people (like me) who start migrating to phoenix (and elixir + erlang).

Thanks,
Thawatchai

**Comment by pma** [comment]
@thawatchai For an alternative monitoring solution, please check this thread https://groups.google.com/forum/#!topic/phoenix-talk/5FCU7ZxkHxI and the blog post by Michael on how to setup exometer: https://medium.com/@mschae/measuring-your-phoenix-app-d63a77b13bda

**Comment by mschae** [comment]
Beat me to it @pma. :)

I highly recommend Exometer. Never managed to get NewRelic to work as they are very secretive about their wire protocol. Plus I didn't want to go down the rabbit hole of nifs just for monitoring (remember my ElixirConf claim: you don't want monitoring to take down your app).

These days I'm happier with the Exometer/StatsD route that I ever was with NewRelic. Just easier to correlate a lot of data. If you have any questions regarding that blog post please feel free to ping me.

**Comment by thawatchai** [comment]
Thanks @pma and @mschae.

Exometer/StatsD looks very interesting. I always want to use something other than New Relic. Probably, this is a chance.

**Comment by thawatchai** [comment]
@mschae, I'm not sure where should I ask a question....

I tried to follow your post and got stuck with this

```
== Compilation error on file lib/plug_exometer.ex ==
** (CompileError) lib/plug_exometer.ex:3: cannot import Plug.register_before_send/2 because it doesn't exist
    (elixir) src/elixir_import.erl:58: :elixir_import.calculate/6
    (elixir) src/elixir_import.erl:20: :elixir_import.import/4
    (stdlib) lists.erl:1352: :lists.mapfoldl/3
    (stdlib) lists.erl:1353: :lists.mapfoldl/3
```

I tried to find a solution, but it seems that Phoenix community is still small. Nobody write a post on how to write a custom plug yet.

Could you give me a hint how to fix this?

BTW, in your post, if you could add the location where we should put each file, it will be really helpful for newbies like me. :smile:

**Comment by pma** [comment]
Replace

import Plug, only: [register_before_send: 2]

with

import Plug.Conn, only: [register_before_send: 2]

and it should work.


On Sat, Jul 11, 2015 at 2:39 PM, Thawatchai Piyawat <
notifications@github.com> wrote:

> @mschae <https://github.com/mschae>, I'm not sure where should I ask a
> question....
>
> I tried to follow your post and got stuck with this
>
> == Compilation error on file lib/plug_exometer.ex ==
> ** (CompileError) lib/plug_exometer.ex:3: cannot import Plug.register_before_send/2 because it doesn't exist
>     (elixir) src/elixir_import.erl:58: :elixir_import.calculate/6
>     (elixir) src/elixir_import.erl:20: :elixir_import.import/4
>     (stdlib) lists.erl:1352: :lists.mapfoldl/3
>     (stdlib) lists.erl:1353: :lists.mapfoldl/3
>
> I tried to find a solution, but it seems that Phoenix community is still
> small. Nobody write a post on how to write a custom plug yet.
>
> Could you give me a hint how to fix this?
>
> BTW, in your post, if you could add the location where we should put each
> file, it will be really helpful for newbies like me. [image: :smile:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phoenixframework/phoenix/issues/402#issuecomment-120619836>
> .
>



-- 
Paulo Almeida
paulo.almeida@timeboxed.co.ao
www.timeboxed.co.ao
T.: (+244) 925721924

**Comment by thawatchai** [comment]
Thanks, @pma. That fixes the error. (and I've just realized that I should try to read the source of `Plug` :flushed:)

Now, I run into this problem in `repo.ex`

```
== Compilation error on file lib/g2k/repo.ex ==
** (CompileError) lib/g2k/repo.ex:7: no super defined for log/2 in module G2k.Repo. Overridable functions available are: log/1
    (elixir) src/elixir_clauses.erl:36: :elixir_clauses.clause/7
    (elixir) src/elixir_def.erl:178: :elixir_def.translate_clause/7
    (elixir) src/elixir_def.erl:167: :elixir_def.translate_definition/8
```

I think that @mschae's post may need to update to reflect the current code of Phoenix. I read `ecto` code now, but my current elixir knowledge isn't enough to fix this issue. :weary: 

BTW, the ability to monitor Phoenix app should be a built-in feature of this framework. I think everybody needs this feature. What do you think?

**Comment by pma** [comment]
I haven't tried this myself but you can check this gist:
https://gist.github.com/romul/4a67f4710d6507f78556 and see if it works.

**Comment by thawatchai** [comment]
Thanks @pma. It works now. :pray:  (but I haven't tried with DataDog yet.)

**Comment by mschae** [comment]
Thanks for answering @pma.

That's correct the signature of the log function was [recently updated](https://github.com/elixir-lang/ecto/blob/master/CHANGELOG.md#v0120).

Thanks for catching those bugs, I'll update the article accordingly.

**Comment by mschae** [comment]
Should be all up to date.

I like my `repo.ex` better :)
https://gist.github.com/mschae/945c7f11423a68010ab2

**Comment by thawatchai** [comment]
Thanks @mschae :+1:

**Comment by thawatchai** [comment]
I have this problem when I try to deploy the code in production.

```
web@e2:~/gotoknow/kv/g2k$ mix compile edown
==> edown (compile)
Compiling src/edown_doclet.erl failed:
src/edown_doclet.erl:116: field packages undefined in record doclet_gen
src/edown_doclet.erl:118: field filemap undefined in record doclet_gen
ERROR: compile failed while processing /media/ebs/v1/sites/gotoknow/kv/g2k/deps/edown: rebar_abort
** (Mix) Could not compile dependency edown, /home/web/.mix/rebar command failed. If you want to recompile this dependency, please run: mix deps.compile edown
```

I installed erlang and elixir from erlang-solutions.com on ubuntu 12.04 (the production server). It even has a newer version of erlang (18). erlang (17) on my Mac installed from homebrew can compile it without a problem.

I try to follow https://github.com/uwiger/unsplit/issues/9 but all erlang-* packages have already been installed.

Could you recommend the machine setting that can make this code work? Thanks for your help. :smile:

**Comment by pap** [comment]
Hi @thawatchai,
I came across a similar thing while using Ubuntu 14.04. Go to the rebar repository and follow the instructions to install it. That should take care of your issue.

**Comment by thawatchai** [comment]
Thanks @pap. :smile:

**Comment by mschae** [comment]
If you're on OTP 18 make sure you're using the latest version of edown. I've updated my post accordingly, make sure you're using `tag: "0.7"`.

**Comment by thawatchai** [comment]
Thanks @mschae. I was trying to add an issue in edown (https://github.com/uwiger/edown/issues/9) but found out later that it's already solved.

**Comment by thawatchai** [comment]
It turned out that I need to delete `mix.lock` to get the new package. Specifying it in `mix.exs` isn't enough. `mix deps.clean --all` then `mix deps.get` still gets the old package.

Did I do anything incorrectly?

**Comment by thawatchai** [comment]
@mschae, after all can be compiled, now it is this :sob:

```
[info] GET /api/posts/adjacent_posts
[debug] Processing by G2k.Api.PostController.adjacent_posts/2
  Parameters: %{"format" => "json", "id" => "3030"}
  Pipelines: [:api]
[info] Sent 500 in 14ms
[error] #PID<0.423.0> running G2k.Endpoint terminated
Server: localhost:3003 (http)
Request: GET /api/posts/adjacent_posts?id=3030
** (exit) an exception was raised:
    ** (ArgumentError) argument error
        (stdlib) :ets.lookup(:exometer_2, [:g2k, :ecto, :query_exec_time])
        src/exometer.erl:213: :exometer.update_/2
        (g2k) lib/g2k/repo.ex:7: G2k.Repo.log/1
        (ecto) lib/ecto/adapters/sql.ex:166: Ecto.Adapters.SQL.query/4
        (ecto) lib/ecto/adapters/sql.ex:390: Ecto.Adapters.SQL.all/6
        (ecto) lib/ecto/repo/queryable.ex:21: Ecto.Repo.Queryable.all/4
        (ecto) lib/ecto/repo/queryable.ex:52: Ecto.Repo.Queryable.one/4
        (g2k) web/controllers/api/post_controller.ex:12: G2k.Api.PostController.adjacent_posts/2
```

It's from this line in `repo.ex`

```
:ok = :exometer.update ~w(g2k ecto query_exec_time)a, (log_entry.query_time + log_entry.queue_time || 0) / 1_000
```

**Comment by pap** [comment]
Have you added exometer to the applications list?

**Comment by mschae** [comment]
Exactly. Also you would've needed to run `mix deps.update edown`.

**Comment by thawatchai** [comment]
after adding `:exometer` to the application list, the error becomes this at the same line of code

```
[info] GET /api/posts/adjacent_posts
[debug] Processing by G2k.Api.PostController.adjacent_posts/2
  Parameters: %{"format" => "json", "id" => "3030"}
  Pipelines: [:api]
[info] Sent 500 in 13ms
[error] #PID<0.362.0> running G2k.Endpoint terminated
Server: localhost:3003 (http)
Request: GET /api/posts/adjacent_posts?id=3030
** (exit) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :not_found}
        (g2k) lib/g2k/repo.ex:7: G2k.Repo.log/1
        (ecto) lib/ecto/adapters/sql.ex:166: Ecto.Adapters.SQL.query/4
        (ecto) lib/ecto/adapters/sql.ex:390: Ecto.Adapters.SQL.all/6
        (ecto) lib/ecto/repo/queryable.ex:21: Ecto.Repo.Queryable.all/4
        (ecto) lib/ecto/repo/queryable.ex:52: Ecto.Repo.Queryable.one/4
        (g2k) web/controllers/api/post_controller.ex:12: G2k.Api.PostController.adjacent_posts/2
        (g2k) web/controllers/api/post_controller.ex:1: G2k.Api.PostController.phoenix_controller_pipeline/2
        (g2k) lib/phoenix/router.ex:281: G2k.Router.dispatch/2
```

Is this because I don't have StatsD (DataDog) run on my development machine (a Mac)? Which parts should I add conditions not to run in dev environment?

**Comment by mschae** [comment]
That's fine you don't need StatsD to run. Well you do if you want to export the numbers somewhere but it won't fail if you don't. :)

Your current error tells me that you either don't have the `conf/exometer.exs` or that you haven't required it in `conf/conf.exs`. If none of that is true please post it in a gist somewhere so we can debug it.

**Comment by mschae** [comment]
In you `config/config.exs` you should have this line:

```Elixir
import_config "exometer.exs"
```

**Comment by thawatchai** [comment]
@mschae It works now. Data from ecto is shown nicely in DataDog.

However, `resp_time` and `resp_count` are all zero. So, I think it's because I don't use `PlugExometer` in my router yet. Anyhow, I can't find a way to use it correctly.

Could you add how to use `PlugExometer` in your post?

**Comment by thawatchai** [comment]
Oh, I've found it in the bottom of this http://www.phoenixframework.org/docs/understanding-plug. (How did I overlook that? :flushed:)

I can just add `plug PlugExometer` in the pipeline, and it works now.

**Comment by thawatchai** [comment]
@mschae, I try to add `app_name.webapp.full_path` report to exometer.

I added the following lines in `PlugExometer`

```
:ok = Exometer.update [:g2k, :webapp, :full_path], Plug.Conn.full_path(conn)
```

Then, the following lines in `exometer_conf.exs`

in `predefined`
```
  {
    [app_name, :webapp, :full_path],
    :spiral,
    []
  },
```

in `report`

```
    {
      :exometer_report_statsd,
      [app_name, :webapp, :full_path], :one, polling_interval, true
    },
```

The stat items are shown in DataDog, but it contains no path. Could you help me to get the stats of request paths? I think full_path is one more thing we need that makes the solution in your post comparable with having New Relic.

Thank you very much. :smile:

**Comment by JKGisMe** [comment]
I tried following this issue and the blog post about setting it up, 

It seems to compile ok but when i try to access a route I get a similar error as @thawatchai had 
```
[info] GET /squares
[debug] Processing by SquareSquareBackend.SquareController.index/2
  Parameters: %{}
  Pipelines: [:api]
[info] Sent 500 in 777ms
[error] #PID<0.477.0> running SquareSquareBackend.Endpoint terminated
Server: localhost:4000 (http)
Request: GET /squares
** (exit) an exception was raised:
    ** (MatchError) no match of right hand side value: {:error, :not_found}
        (square_square_backend) lib/square_square_backend/repo.ex:6: SquareSquareBackend.Repo.log/1
        (ecto) lib/ecto/adapters/sql.ex:211: Ecto.Adapters.SQL.query/5
        (ecto) lib/ecto/adapters/sql.ex:170: Ecto.Adapters.SQL.query!/5
        (ecto) lib/ecto/adapters/sql.ex:465: Ecto.Adapters.SQL.execute/6
        (ecto) lib/ecto/repo/queryable.ex:95: Ecto.Repo.Queryable.execute/5
        (ecto) lib/ecto/repo/queryable.ex:15: Ecto.Repo.Queryable.all/4
        (square_square_backend) web/controllers/square_controller.ex:9: SquareSquareBackend.SquareController.index/2
        (square_square_backend) web/controllers/square_controller.ex:1: SquareSquareBackend.SquareController.phoenix_controller_pipeline/2
        (square_square_backend) lib/phoenix/router.ex:265: SquareSquareBackend.Router.dispatch/2
        (square_square_backend) web/router.ex:3: SquareSquareBackend.Router.do_call/2
        (square_square_backend) lib/square_square_backend/endpoint.ex:1: SquareSquareBackend.Endpoint.phoenix_pipeline/1
```
I did add the `import_config "exometer.exs"` to my `conf/conf.ex` so I'm not sure what's going on. I'm new to elixir/phoenix as well. I was a little confused about the `morgue` and `:morgue` bits. I put my app name there instead but I wasn't sure. 

I have everything here: https://github.com/JKGisMe/square_square_backend/tree/addExometer

Any advice or tips would be greatly appreciated.

**Comment by mschae** [comment]
@JKGisMe I am sorry you were right to put you app name there instead of `morgue` - that's the name of our app I stole the gist from. :) I updated the gist accordingly: https://gist.github.com/mschae/945c7f11423a68010ab2

Your error tells me that exometer couldn't find a predefined log entry for what you were trying to do. Specifically line 6 meaning `query_queue_time` wasn't defined. I checked the provided exometer.conf from my article and it's indeed missing. Sorry. Updated: https://gist.github.com/mschae/3c1d7fc77613ada7f59e/revisions

Let me know if adding that piece to your `exometer.conf` fixes it.

@thawatchai you are trying to provide a literal value (the full path) to a function that expects numerical values (`Exometer.update`). You can't just pass a string to a counter and expect it to understand that you are trying to record the number of individual values here. Instead you'd have to define new metrics for every full path and then increment the counter.

Recording path-specific metrics is not that trivial.

**Comment by JKGisMe** [comment]
Thanks so much! That worked for everything except when I would create a new record but I think I fixed it. The `queue_time` would be nil so it would error when it tried to add nil to the `query_time`. Just added a set of parenthesis in `repo.ex` and everything seems to be working!  I changed 
```
:ok = :exometer.update ~w(square_square_backend ecto query_exec_time)a, (log_entry.query_time + log_entry.queue_time || 0) / 1_000
```
to
```
:ok = :exometer.update ~w(square_square_backend ecto query_exec_time)a, (log_entry.query_time + (log_entry.queue_time || 0)) / 1_000
```

**Comment by chrismccord** [comment]
@JKGisMe glad things are moving in the right direction for you, but can you take the discussion to the ML or relevant repo? Thanks!

**Comment by JKGisMe** [comment]
The comments here were indispensable to me. I'm not familiar with "ML"

**Comment by lancehalvorsen** [comment]
"ML" refers to the phoenix-talk mailing list:
https://groups.google.com/forum/#!forum/phoenix-talk

**Comment by JKGisMe** [comment]
Ah, ok. I find that to be really user unfriendly. I won't come back here. 

Just saying... don't single out & blame the last person who happened to reply to annoying issue. Everyone up the list contributed.

**Comment by mschae** [comment]
@JKGisMe it's not your fault and I think it wasn't @chrismccord's intention to single you out. He is super eager to help. But he is correct in saying that this is not something we should discuss via GH issues.

This is more my fault than anyone else's I should've taken it to the list. I apologise.

Please do come back here whenever there is something specific to the framework. Everything else should indeed be discussed on the list so a broader audience can find it and we keep this repo clean.

Again, he wasn't singling you out you just happened to be the last one to comment here.

**--- For everyone ---**

Let's take this here:
https://groups.google.com/forum/#!searchin/phoenix-talk/monitoring/phoenix-talk/5FCU7ZxkHxI/kSKjiTgJzAAJ

**Comment by chrismccord** [comment]
> Ah, ok. I find that to be really user unfriendly. I won't come back here.

@JKGisMe My apologies if I came off anything other than friendly. We try to keep issues related to phoenix core and other discussions on the mailing list, see https://github.com/phoenixframework/phoenix/blob/master/CONTRIBUTING.md for more details.

I'm working through 300 emails in my inbox following my vacation, so you happened to be the last one in this thread. I did not mean to single you out. Really, I hope you stick around and feel welcome to contributing and asking for help, it would pain me otherwise to see that not be the case :(
